src/plugins/shell/include/plugins/shell/shell.h

```diff
@@ -0,0 +1,124 @@
+#pragma once
+
+#include <functional>
+#include <memory>
+#include <string>
+#include <array>
+#include <limits>
+
+#include "plugin_base.h"
+
+namespace mavsdk {
+
+class ShellImpl;
+class System;
+
+/**
+ * @brief This class allow users to communicate with vehicle's system shell.
+ */
+class Shell : public PluginBase {
+public:
+    /**
+     * @brief Constructor. Creates the plugin for a specific System.
+     *
+     * The plugin is typically created as shown below:
+     *
+     *     ```cpp
+     *     auto shell = std::make_shared<Shell>(system);
+     *     ```
+     *
+     * @param system The specific system associated with this plugin.
+     */
+    explicit Shell(System& system);
+
+    /**
+     * @brief Destructor (internal use only).
+     */
+    ~Shell();
+
+    /**
+     * @brief Shell Message type
+     */
+    struct ShellMessage {
+        bool need_response; /**< @brief Set if wants the receiver to send a response. */
+        uint16_t timeout; /**< @brief Timeout for reply data in ms. */
+        std::string data; /**< @brief Serial data. */
+    };
+
+    /**
+     * @brief Shell requests results type.
+     */
+    struct Result {
```

Comment:
`Result` is usually an `enum` so this is bit different then other plugins.

What about using this callback to return:
```
typedef std::function<void(Result result, ShellMessage response)> result_callback_t;
```

I'm not sure if it's nicer but it's more consistent :thinking:?
@JonasVautherin do you have an opinion?

Reply:
This does not allow [this feature](https://github.com/mavlink/MAVSDK/pull/878#discussion_r335893631). I should change it corresponding other modules.

Reply:
[Done](https://github.com/mavlink/MAVSDK/pull/878/commits/d985378c1e4bb4ef93d9fa02d7f6352916d8158e#diff-f4c5701958a5b01cfeef470e26440de5R49)

Reply:
[Done](https://github.com/mavlink/MAVSDK/pull/878/commits/d985378c1e4bb4ef93d9fa02d7f6352916d8158e#diff-8ca1b5f6cd83dbe50248a5028f19845eR78)